File: journal/last_check.go
package main

import (
	"encoding/json"
	"net/http"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// last_check.go — the SERVER-SIDE "when did you last look at what changed" boundary.
// Wave 5 Lane 5B, extending §2.3's per-user event state.
//
// WHY IT MOVED HERE. Wave 4 Lane 4A put it in `localStorage`
// (`web/src/components/events/lastCheck.js`) and said in the same breath that this was a limitation
// rather than a design: the boundary does not survive a new device, a cleared profile or a second
// browser, so "since your last check" silently becomes "in the last 24 hours" for a returning user
// on a second machine. 4A's handoff, GAPS.md and the Wave 4 integration report all recorded the
// same thing — the server-side home is journal's `user_event_state` (§2.3), and moving it is a
// CONTRACT AMENDMENT rather than a lane decision. 5B lands `/api/changed`, which is the surface
// that reads the boundary, so this is the wave that owes the move.
//
// WHAT IT IS, EXACTLY: one unix second per user. Not per event, not per ticker, not per surface.
// It lives beside `event_state.json` in the same partition, so D-06 still holds — `rm -rf
// {TRADES_DIR}/{uid}/` removes it with everything else of theirs.
//
// THE COPY RULE SURVIVES THE MOVE, AND THAT IS THE POINT. A default 24-hour window may NEVER be
// labelled "since your last check" (Wave 4 Lane 4A's locked decision 6). Server-side, that means:
//
//   - `GET` returns `{"lastCheck": null}` when the user has never recorded one. NULL, not "now",
//     not "24 hours ago" — a boundary the server invented would be indistinguishable at the client
//     from one the user earned, and the heading would then lie.
//   - The client omits `since` when it is null, and `/api/changed` answers `basis: "default24h"`,
//     which is the field the heading turns on.
//
// IT IS NEVER ADVANCED BY A READ. `GET` does not stamp; only an explicit `POST` does, and the
// client sends it only once the surface has actually rendered an answer. Advancing the boundary on
// a read would silently skip past events the user never saw, which is the one failure the whole
// object exists to prevent.

const lastCheckFile = "last_check.json"

// lastCheckRecord is deliberately a struct rather than a bare int, so a later field (a per-surface
// boundary, say) is an additive change rather than a format migration.
type lastCheckRecord struct {
	At        int64 `json:"at"`
	UpdatedAt int64 `json:"updatedAt"`
}

type lastCheckStore struct {
	base string
	mu   sync.Mutex
	docs *documentRepository
}

func openLastCheckStore(base string) (*lastCheckStore, error) {
	return openLastCheckStoreWithRepository(base, nil)
}

func openLastCheckStoreWithRepository(base string, docs *documentRepository) (*lastCheckStore, error) {
	if err := os.MkdirAll(base, 0o755); err != nil {
		return nil, err
	}
	return &lastCheckStore{base: base, docs: docs}, nil
}

func (s *lastCheckStore) path(uid string) string {
	return filepath.Join(s.base, uid, lastCheckFile)
}

// get returns `(at, ok)`. `ok` is false when the user has never recorded a boundary — the state the
// copy rule turns on, and the reason this is not a bare `int64` return.
func (s *lastCheckStore) get(uid string) (int64, bool) {
	if reservedUID(uid) {
		return 0, false
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	var rec lastCheckRecord
	var b []byte
	var err error
	if s.docs != nil {
		b, _, err = s.docs.load(uid, "last_check", s.path(uid))
	} else {
		b, err = os.ReadFile(s.path(uid))
	}
	if err != nil || json.Unmarshal(b, &rec) != nil || rec.At <= 0 {
		return 0, false
	}
	return rec.At, true
}

// set records the boundary. MONOTONIC: a later stamp wins and an earlier one is ignored rather than
// applied. Two tabs racing must not move a user's reading position BACKWARDS — that would re-show
// events they had already been shown, which erodes the same trust as skipping them.
func (s *lastCheckStore) set(uid string, at int64) (int64, error) {
	if reservedUID(uid) {
		return 0, errSubscriptionReservedUID
	}
	s.mu.Lock()
	defer s.mu.Unlock()

	current := lastCheckRecord{}
	var b []byte
	var loadErr error
	if s.docs != nil {
		b, _, loadErr = s.docs.load(uid, "last_check", s.path(uid))
	} else {
		b, loadErr = os.ReadFile(s.path(uid))
	}
	if loadErr == nil {
		_ = json.Unmarshal(b, &current)
	}
	if at <= current.At {
		return current.At, nil
	}
	next := lastCheckRecord{At: at, UpdatedAt: time.Now().Unix()}
	if s.docs != nil {
		if err := s.docs.save(uid, "last_check", next); err != nil {
			return current.At, err
		}
		return next.At, nil
	}
	if err := os.MkdirAll(filepath.Join(s.base, uid), 0o755); err != nil {
		return current.At, err
	}
	b, err := json.MarshalIndent(next, "", "  ")
	if err != nil {
		return current.At, err
	}
	tmp := s.path(uid) + ".tmp"
	if err := os.WriteFile(tmp, b, 0o644); err != nil {
		return current.At, err
	}
	if err := os.Rename(tmp, s.path(uid)); err != nil {
		return current.At, err
	}
	return next.At, nil
}

// ---- HTTP surface ----

// init attaches the pair through the same Wave 0 seam `event_state.go` uses.
func init() {
	registerSubscriptionRoute(func(s *Server, mux *http.ServeMux) {
		s.registerLastCheckAPI(mux)
	})
}

func (s *Server) registerLastCheckAPI(mux *http.ServeMux) {
	store, err := openLastCheckStoreWithRepository(s.cfg.TradesDir, s.documents)
	if err != nil {
		// Additive: the journal must still serve everything else. The routes are simply not
		// mounted and the client falls back to its local boundary, which is the pre-Wave-5
		// behaviour and is honest about itself.
		return
	}

	// Both require a session. A guest has no partition, and a guest boundary is exactly what
	// `localStorage` is still for.
	mux.HandleFunc("GET /event-state/last-check", s.requireAuth(func(w http.ResponseWriter, r *http.Request) {
		at, ok := store.get(userID(r))
		if !ok {
			// NULL, not a substituted default. See the header: a server-invented boundary would be
			// indistinguishable at the client from one the user earned.
			writeJSON(w, http.StatusOK, map[string]any{"lastCheck": nil})
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"lastCheck": at})
	}))

	mux.HandleFunc("POST /event-state/last-check", s.requireAuth(func(w http.ResponseWriter, r *http.Request) {
		var body struct {
			At *int64 `json:"at"`
		}
		_ = json.NewDecoder(r.Body).Decode(&body)
		at := time.Now().Unix()
		if body.At != nil && *body.At > 0 {
			at = *body.At
		}
		stored, err := store.set(userID(r), at)
		if err != nil {
			writeJSON(w, http.StatusBadRequest, map[string]any{
				"error": err.Error(), "code": "last_check_unwritable"})
			return
		}
		writeJSON(w, http.StatusOK, map[string]any{"lastCheck": stored})
	}))
}

File: web/src/components/events/lastCheck.js
// lastCheck — "when did you last look at what changed".
//
// WAVE 5B MOVED THIS TO THE SERVER, and this file is now the client half of a two-tier boundary
// rather than the whole thing.
//
//   signed in   → journal's per-user partition, via `GET/POST /api/event-state/last-check`
//                 (contract §2.3). Survives a new device, a cleared profile and a second browser.
//   signed out  → `localStorage`, exactly as before. A guest has no partition, and a guest boundary
//                 is precisely what local storage is for.
//
// Wave 4 Lane 4A shipped the local half and said in the same breath that it was a limitation rather
// than a design: "It does not survive a new device, a cleared profile or a second browser." 4A's
// handoff, GAPS.md and the Wave 4 integration report all named journal's `user_event_state` as the
// server-side home and recorded the move as a CONTRACT AMENDMENT. 5B lands `/api/changed`, which is
// the surface that reads the boundary, so this is the wave that owes the move.
//
// THE COPY RULE THIS FILE EXISTS TO ENFORCE — unchanged, and it is a rule, not a preference: a
// default 24-hour window may NEVER be labelled "since your last check". No stored timestamp ⇒ the
// caller omits `since`, the server answers `since.basis === "default24h"`, and the heading reads
// "IN THE LAST 24 HOURS". `hasLastCheck` and the resolved `basis` are how a caller knows which
// sentence it is allowed to write.
//
// THE LOCAL COPY IS STILL WRITTEN FOR A SIGNED-IN USER, deliberately. It is the offline floor: if
// journal is unreachable the surface still says something true about this browser rather than
// silently regressing to the 24-hour window. The SERVER value wins whenever both exist — the local
// one can only be older or equal, since every stamp goes to both.

const KEY = "attestel.lastCheck";
const BASE = import.meta.env.VITE_GATEWAY_URL || "";

// Keyed by uid so two accounts on one browser do not inherit each other's boundary. Signed-out is
// the literal "guest" — a real, separate reading position, not a missing one.
const slot = (uid) => `${KEY}.${uid || "guest"}`;

// readLastCheck returns unix SECONDS from LOCAL storage, or 0 when there is none. 0 is falsy on
// purpose: every caller then omits `since` and falls back to the 24-hour heading without a second
// branch. It stays synchronous because a surface has to render its first frame before any fetch
// resolves, and the local value is the honest thing to render meanwhile.
export function readLastCheck(uid) {
  try {
    const raw = window.localStorage.getItem(slot(uid));
    const n = Number(raw);
    return Number.isFinite(n) && n > 0 ? n : 0;
  } catch {
    // Private mode, disabled storage, quota. A missing boundary is a fully supported state, so
    // this degrades to "the last 24 hours" rather than surfacing an error at the user.
    return 0;
  }
}

// writeLastCheck is called ONLY when the surface was actually viewed — not on mount of a parent,
// not on a route change that never painted it. Writing it earlier would silently advance the
// boundary past events the user never saw, which is the one failure this whole object exists to
// prevent.
export function writeLastCheck(uid, at = Math.floor(Date.now() / 1000)) {
  try {
    window.localStorage.setItem(slot(uid), String(at));
  } catch {
    /* storage unavailable — the surface stays on the 24-hour window, which is honest */
  }
}

export const hasLastCheck = (uid) => readLastCheck(uid) > 0;

// resolveLastCheck is the SERVER-FIRST read. It returns `{ at, source }` where `source` is
// "server" | "local" | "none", so a caller can tell a boundary that will follow the user to their
// next device from one that will not.
//
// A guest never calls the server: there is no partition to read and the route is a 401 by design.
// A signed-in user whose journal is unreachable falls back to the local value — degraded, but
// truthful, and never to an invented "now".
export async function resolveLastCheck(uid) {
  const local = readLastCheck(uid);
  if (!uid) return { at: local, source: local ? "local" : "none" };

  try {
    const res = await fetch(`${BASE}/api/event-state/last-check`, { credentials: "include" });
    if (res.ok) {
      const body = await res.json();
      const at = Number(body?.lastCheck);
      // `null` is a REAL ANSWER — "this user has never recorded a boundary" — and it is not the
      // same as a failed read. Both fall through to the local value, but only the failed one is
      // a degradation, and neither may become an invented timestamp.
      if (Number.isFinite(at) && at > 0) {
        // Keep the local copy in step so the offline floor is never behind the server.
        if (at > local) writeLastCheck(uid, at);
        return { at, source: "server" };
      }
      return { at: local, source: local ? "local" : "none" };
    }
  } catch {
    /* offline or journal down — fall through to the local value */
  }
  return { at: local, source: local ? "local" : "none" };
}

// recordLastCheck stamps BOTH tiers. Local first and unconditionally, so the boundary survives even
// if the request fails; then the server, whose store is monotonic and will ignore a stamp older
// than one it already holds.
//
// Same rule as `writeLastCheck`: call it only once the surface has actually rendered an answer.
export async function recordLastCheck(uid, at = Math.floor(Date.now() / 1000)) {
  writeLastCheck(uid, at);
  if (!uid) return;
  try {
    await fetch(`${BASE}/api/event-state/last-check`, {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ at }),
    });
  } catch {
    /* the local stamp stands; the next successful call will carry it forward */
  }
}
